src/coreComponents/constitutive/fluid/BlackOilFluid.hpp

```diff
@@ -19,57 +19,1281 @@
+                                         arraySlice3d< real64, multifluid::USD_PHASE_COMP_DC - 2 > const & dPhaseCompFraction_dGlobalCompFraction ) const
+{
+  using PT = BlackOilFluid::PhaseType;
+
+  localIndex const ipOil   = m_phaseOrder[PT::OIL];
+  localIndex const ipGas   = m_phaseOrder[PT::GAS];
+  localIndex const ipWater = m_phaseOrder[PT::WATER];
+
+  localIndex const icOil   = ipOil;
+  localIndex const icGas   = ipGas;
+  localIndex const icWater = ipWater;
+
+  real64 const zo = composition[icOil];
+  real64 const zg = composition[icGas];
+  real64 const zw = composition[icWater];
+
+  // 1. Make everything zero first
+
+  for( localIndex ip = 0; ip < NP_BO; ++ip )
+  {
+    phaseFraction[ip] = 0.;
+    if( needDerivs )
+    {
+      dPhaseFraction_dPressure[ip] = 0.0;
+    }
+    for( localIndex ic = 0; ic < NC_BO; ++ic )
+    {
+      phaseCompFraction[ip][ic] = 0.0;
+      if( needDerivs )
+      {
+        dPhaseFraction_dGlobalCompFraction[ip][ic] = 0.;
+        dPhaseCompFraction_dPressure[ip][ic] = 0.0;
+        for( localIndex jc = 0; jc < NC_BO; ++jc )
+        {
+          dPhaseCompFraction_dGlobalCompFraction[ip][ic][jc] = 0.0;
+        }
+      }
+    }
+  }
+
+  // 2. Check feed first, and if only water is present (e.g., water inj), then skip
+
+  if( zw >= 1. - minForPhasePresence )
+  {
+    phaseFraction[ipWater] = zw;
+    if( needDerivs )
+    {
+      dPhaseFraction_dGlobalCompFraction[ipWater][icWater] = 1.;
+    }
+    phaseCompFraction[ipWater][icWater] = 1.0;
+    return;
+  }
+
+  // 3. Compute Rs (saturated) as a function of pressure
+
+  // oil
+  real64 const & oilSurfaceMoleDensity = m_PVTOView.m_surfaceMoleDensity[PT::OIL];
+  real64 RsSat = 0.0;
+  real64 dRsSat_dP = 0.0;
+  computeRs( pressure, RsSat, dRsSat_dP );
+
+  // gas
+  real64 const & gasSurfaceMoleDensity = m_PVTOView.m_surfaceMoleDensity[PT::GAS];
+
+  if( RsSat < minForPhasePresence )
+  {
+    RsSat = minForPhasePresence;
+  }
+
+  // 4. Use the saturated Rs and density to compute the gas phase fraction
+  //    Note : we assume the oil component cannot enter the gas phase
+
+  real64 const Kg = ( oilSurfaceMoleDensity + gasSurfaceMoleDensity * RsSat ) / ( RsSat * gasSurfaceMoleDensity );
+  real64 const dKg_dP = -oilSurfaceMoleDensity / gasSurfaceMoleDensity * dRsSat_dP / ( RsSat * RsSat );
+  real64 const gasPhaseFraction = zo / ( 1. - Kg ) + zg;
+  real64 const dGasPhaseFraction_dP = zo / ( ( 1. - Kg ) * ( 1. - Kg ) ) *  dKg_dP;
+  real64 const dGasPhaseFraction_dzo = 1. / ( 1. - Kg );
+  real64 const dGasPhaseFraction_dzg = 1.;
+
+  // 4. Update phase fraction and phase component fractions
+
+  // 4.1 The gas phase is present
+  if( ( gasPhaseFraction > 0 ) && ( gasPhaseFraction < 1 ) )
+  {
+
+    // phase fractions
+    phaseFraction[PT::OIL] = 1. - gasPhaseFraction - zw;
+    phaseFraction[PT::GAS] = gasPhaseFraction;
+    phaseFraction[PT::WATER] =  zw;
```

Comment:
Why is `phaseFraction` not using the same phase ordering as other output quantities (i.e. indexing though `ipGas`, `ipOil`, `ipWater`)?